[VP9e] Plugin interface replaced by the core interface

DmitryMalishev opened the pull request:
[1] The change was checked on Windows for no regressions.
[2] For Linux only compiling and building possibility was checked.

artem-shaporenko commented on _studio/mfx_lib/encode_hw/vp9/include/mfx_vp9_encode_hw_utils.h:
```diff
@@ -512,7 +513,7 @@ template <typename T> mfxStatus RemoveExtBuffer(T & par, mfxU32 id)
         FrameLocker & operator =(FrameLocker const &);
 
-        mfxCoreInterface * m_core;
+        VideoCORE * m_core;
```
please fix alignment

artem-shaporenko commented on _studio/mfx_lib/encode_hw/vp9/include/mfx_vp9_encode_hw_utils.h:
```diff
@@ -589,7 +590,7 @@ constexpr auto NUM_OF_SUPPORTED_EXT_BUFFERS = 7; // mfxExtVP9Param, mfxExtOpaque
                         sFrameEx *pRecFrame,
                         std::vector<sFrameEx*>&dpb,
-                        mfxCoreInterface *pCore);
+        VideoCORE *pCore);
```
the same, alignment

mgonchar commented on _studio/mfx_lib/encode_hw/vp9/src/mfx_vp9_encode_hw.cpp:
```diff
@@ -37,6 +37,8 @@ mfxStatus MFXVideoENCODEVP9_HW::Query(mfxCoreInterface *core, mfxVideoParam *in,
     }
     else
     {
+        core->GetHWType();
```
you forgot to assign hw type to some variable?

dvrogozh replied:
yeah, this one looks weird and needs to be fixed

DmitryMalishev replied:
This call is done this way on purpose. During this call MSDK Core is being initialized that is the mandatory step to keep working. Feel free to suggest an alternative!

artem-shaporenko replied:
That should not go this way, core initialized during MFXInit, but not by  GetHWType, if it is working this way - you are doing something wrong.....

mgonchar commented on _studio/mfx_lib/encode_hw/vp9/src/mfx_vp9_encode_hw.cpp:
```diff
@@ -198,6 +196,8 @@ mfxStatus MFXVideoENCODEVP9_HW::Init(mfxVideoParam *par)
 
     m_video = *par;
 
+    m_pCore->GetHWType();
```
same

DmitryMalishev replied:
This call is done this way on purpose. During this call MSDK Core is being initialized that is the mandatory step to keep working. Feel free to suggest an alternative!

mgonchar commented on _studio/mfx_lib/encode_hw/vp9/src/mfx_vp9_encode_hw_utils.cpp:
```diff
@@ -838,20 +830,46 @@ mfxStatus CopyRawSurfaceToVideoMemory(
+
+    if (iopattern == MFX_IOPATTERN_IN_SYSTEM_MEMORY ||
+        (iopattern == MFX_IOPATTERN_IN_OPAQUE_MEMORY && (opaq_type & MFX_MEMTYPE_SYSTEM_MEMORY)))
+        sts = core.GetFrameHDL(mid, handle);
```
I think you can get rid of sts variable at all and just simply return at each if statement

dvrogozh commented:
@lizhong1008 : that's a patch which should help enable VP9e in ffmpeg-qsv.

DmitryMalishev commented:
I removed core->GetHWType() calls (not required for Linux) and fixed other requests.

Miroslav, I left GetNativeHandleToRawSurface() as it is, because I'm considering single Return at the end is better for debugging.

dvrogozh commented:
@mgonchar , @artem-shaporenko : please, review again.

DmitryMalishev commented:
Rebase is done.

dvrogozh commented:
@mgonchar. @artem-shaporenko : can you review, please.

mgonchar commented on _studio/mfx_lib/encode_hw/vp9/src/mfx_vp9_encode_hw.cpp:
```diff
@@ -107,16 +104,13 @@ mfxStatus MFXVideoENCODEVP9_HW::QueryIOSurf(mfxCoreInterface *core, mfxVideoPara
     ENCODE_CAPS_VP9 caps = {};
-    mfxPlatform platform = {};
-    MFX_CHECK(MFX_ERR_NONE == QueryCapsAndPlatform(&_core, caps, platform, GetGuid(toValidate), par->mfx.FrameInfo.Width, par->mfx.FrameInfo.Height), MFX_ERR_UNSUPPORTED);
+    MFX_CHECK(MFX_ERR_NONE == QueryCaps(core, caps, GetGuid(toValidate), par->mfx.FrameInfo.Width, par->mfx.FrameInfo.Height), MFX_ERR_UNSUPPORTED);
```
I suggest to get the function call out of macro. During the macro expansion at debug current approach will cause function to be called twice

DmitryMalishev replied:
Good point! Despite it is not reproduced for this particular case, I'd better move the function out of the macros to prevent possible issue in future.

dvrogozh replied:
Not addressed.

mgonchar commented on _studio/mfx_lib/encode_hw/vp9/src/mfx_vp9_encode_hw.cpp:
```diff
@@ -237,16 +229,19 @@ mfxStatus MFXVideoENCODEVP9_HW::Init(mfxVideoParam *par)
         {
-            request.NumFrameMin = request.NumFrameSuggested = opaq.In.NumSurface;
+            request.Type = MFX_MEMTYPE_D3D_INT;
+            request.NumFrameMin = opaq.In.NumSurface;
```
shouldn't we set NumFrameSuggested ?

DmitryMalishev replied:
'NumFrameSuggested' not used for allocation if fact. I'm removing NumFrameSuggested for all cases here to make it more clear.

dvrogozh commented:
@DmitryMalishev : can you, please, rebase? There is a conflict in VP9 files.

DmitryMalishev commented:
A added a change to the recent request for the macros, plus rebased the commit.

mgonchar commented on _studio/mfx_lib/encode_hw/vp9/include/mfx_vp9_encode_hw_utils.h:
```diff
@@ -216,16 +216,16 @@ enum // identifies memory type at encoder input w/o any details
         {
-            sts = pCore->DecreaseReference(pCore->pthis, &pFrame->pSurface->Data);
+            sts = pCore->DecreaseReference(&pFrame->pSurface->Data);
             pFrame = 0;
```
nullptr
